src/main/webapp/app/exercises/programming/shared/code-editor/ace/code-editor-ace.component.ts

```diff
@@ -492,42 +488,70 @@ export class CodeEditorAceComponent implements AfterViewInit, OnChanges, OnDestr
      * @param line line of code where the feedback inline component will be added to.
      */
     addLineWidgetWithFeedback(line: number) {
-        // Create new feedback element from the DOM
-        this.linesWithNewFeedback = [...this.linesWithNewFeedback, line];
-        // Update DOM so that the new feedback DOM node exists
+        // Get the feedback DOM node
+        const inlineFeedbackNode = this.getInlineFeedbackNode(line);
+        if (!inlineFeedbackNode) {
+            throw new Error("Couldn't find inline feedback node for line " + line + '. Make sure that it is rendered by Angular.');
+        }
+        const lineWidget = {
+            row: line,
+            fixedWidth: true,
+            coverGutter: true,
+            el: inlineFeedbackNode,
+        };
+        const displayedWidget = (this.editorSession.lineWidgets ?? [])[line];
+        if (!displayedWidget) {
+            this.editorSession.widgetManager.addLineWidget(lineWidget);
+            lineWidget.el.querySelector('textarea')?.focus();
+        }
+    }
+
+    /**
+     * Add lineWidget for specific line of code.
+     * (ACE line widget, needs to be manually handled because ACE is not well integrated into Angular)
+     * @param line line of code where the feedback inline component will be added to.
+     */
+    removeLineWidget(line: number) {
+        const widget = (this.editorSession.lineWidgets ?? [])[line];
+        if (widget) {
+            this.editorSession.widgetManager.removeLineWidget(widget);
+        }
+    }
+
+    /**
+     * Sync the line widgets of the ACE editor to the current feedbacks.
+     */
+    async updateLineWidgets() {
+        // Make Angular update the DOM
         this.changeDetectorRef.detectChanges();
-        const inlineFeedback = this.getInlineFeedbackNode(line);
-        if (inlineFeedback) {
-            const lineWidget = {
-                row: line,
-                fixedWidth: true,
-                coverGutter: true,
-                el: inlineFeedback,
-            };
-            // Check if lineWidget is already displayed
-            if (this.editorSession.lineWidgets) {
-                const displayedWidget = this.editorSession.lineWidgets.find((w: any) => w && w.row === lineWidget.row);
-                if (!displayedWidget) {
-                    this.editorSession.widgetManager.addLineWidget(lineWidget);
-                    lineWidget.el.querySelector('textarea')?.focus();
-                }
-            } else {
-                this.editorSession.widgetManager.addLineWidget(lineWidget);
-                lineWidget.el.querySelector('textarea')?.focus();
-            }
+        // Wait for one render cycle for all DOM updates to be done
+        await new Promise((resolve) => setTimeout(resolve, 0));
+        // Remove all existing line widgets
+        const linesWithWidgets = this.editorSession.lineWidgets?.map((l: any) => l?.row).filter((row: number | undefined) => row != null) ?? [];
```

Comment:
`row: number | undefined` but then checking for `null`?

Reply:
It's a habit I developed to avoid issues with `null` vs. `undefined`: When you check `x == null` with only two equals signs in JS (or, like here, the opposite `!= null`), it is true for both `null` and `undefined` - so you don't have to worry about that.

However, as this violates our client style guides and it's a bit confusing, I changed it to `undefined` :+1: